refactor: Restructuring of classes to reduce duplication and increase readibility

holtskinner opened the pull request:
- Changed `vision_helpers.py` to input `Document.Page` to reduce extra imports.
- Rewrote Entity image extraction to avoid storing the Pillow image in memory.
- Added error checking for missing shards.
- Added sorting of shards by index in case they are read out of order.
- Fixed bug where `Entity` page numbers would be incorrect in multi-shard documents.
- Simplified creation of `Page` sub classes by using `__post_init__()`
- Renamed `Page.text` and `Table.text` to `document_text` to be more descriptive and match the other `Page` elements.
- Rewrote `table_sample.py` to use more accurate/descriptive variable names.
- Minor formatting of comments/docs

holtskinner committed:
refactor: Change classes to avoid extra copies

holtskinner committed:
Fix `default_factory` error

holtskinner committed:
Update to fix tests

holtskinner committed:
docs: Update Doc Strings

holtskinner committed:
Changed other `documentai_` fields to InitVar
- Fixed table sample

dizcology commented on google/cloud/documentai_toolbox/wrappers/document.py:
```diff
@@ -717,15 +695,34 @@ def split_pdf(self, pdf_path: str, output_path: str) -> List[str]:
+
+        for shard in self.shards:
+            for docai_page in shard.pages:
+                page_vision_annotation = _convert_document_page(
```
Can we take this opportunity to import module instead of functions?  In particular this would become `vision_helpers._convert_document_page`, making its use cases more obvious.

holtskinner replied:
I'm good with that. This was just moving the code that was already present.

dizcology commented on google/cloud/documentai_toolbox/wrappers/document.py:
```diff
@@ -64,17 +62,19 @@ def _entities_from_shards(
-
-    if len(result) > 1 and result[0].documentai_entity.id:
-        result.sort(key=lambda x: int(x.documentai_entity.id))
+        shard_index = int(shard.shard_info.shard_index)
```
Do we know that `shard.shard_info.shard_index` will always be the same as the index of `shard` as it appears in the list `shards`?

In `export_images` below we have the following code:
```
image = entity.crop_image(
                documentai_document=self.shards[entity.shard_index]
            )
```
which suggests that these two notion of "shard index" are indeed the same.  Can we confirm that to be the case?

Alternatively, from the perspective of the Toolbox library, the "index of a shard as it appears in the list shards" is the one actually used, so perhaps we can simply iterate with `for shard_index, shard in enumerate(shards): ...` here instead.

holtskinner replied:
It should be the same, unless there are any missing (which is a bigger issue tbh)

Maybe I should add an exception for missing shards, because that would cause bigger issues down the line

dizcology commented on google/cloud/documentai_toolbox/wrappers/entity.py:
```diff
@@ -26,44 +27,73 @@
+    """Represents a wrapped `documentai.Document.Entity`.
 
     Attributes:
+        shard_index (int):
```
Since these wrapper classes are "public", let's acknowledge that this change of adding a new required argument (and in other places) is a breaking change in the commit message.

holtskinner replied:
No longer applicable

dizcology commented on google/cloud/documentai_toolbox/wrappers/entity.py:
```diff
@@ -26,44 +27,73 @@
-    documentai_entity: documentai.Document.Entity = dataclasses.field(repr=False)
+
+    shard_index: int = dataclasses.field()
+    documentai_entity: dataclasses.InitVar[documentai.Document.Entity]
```
Just to confirm, are we changing all the existing wrapped documentai messages to `InitVar`, and extract information from the those messages in `__post_init__`, but never keeping the whole documentai message as a field?

(Indeed, we should keep this consistent across all wrapper classes - although technically they are no longer "wrappers" in the usual sense.)

holtskinner replied:
That is true. I am changing the documentai messages to InitVar for all wrappers except the `Document` wrapper. The `documentai_document` contains all of the information used to populate the other wrappers.

dizcology commented on google/cloud/documentai_toolbox/wrappers/entity.py:
```diff
@@ -26,44 +27,73 @@
+                    0
+                ].bounding_poly.normalized_vertices
+
+    def crop_image(self, documentai_document: documentai.Document) -> Image.Image:
```
Update the return type to `Optional[Image.Image]` since the method could return `None`.

dizcology commented on google/cloud/documentai_toolbox/wrappers/page.py:
```diff
@@ -161,7 +161,7 @@ class Line:
 
-    documentai_line: documentai.Document.Page.Line
+    documentai_line: dataclasses.InitVar[documentai.Document.Page.Line]
     text: str
```
It feels like we only need the `InitVar` `documentai_line` here, and we can extract the text and store that in a `Line.text` field in `__post_init__` (which has access to `documentai_line`)?

That is, perhaps the `text` field should be a dataclass field with `init=False`, so the caller does not have to pass in both `documentai_line` and `text`?

(Similar situation in some other classes that currently do not have `__post_init__`.)

holtskinner replied:
I agree

dizcology commented on google/cloud/documentai_toolbox/wrappers/page.py:
```diff
@@ -365,10 +365,15 @@ class Page:
+        text (str):
+            Required. The full text of the `Document` containing the `Page`.
+        page_number (int):
+            Required. The page number of the `Page`.
```
Maybe add to the docstring that `page_number` is always page index+1.  But then how is this field used?  Wouldn't `page_index` be more useful?

(In `samples/snippets/table_sample.py` below, for example, the name `page_number` indeed refers to page index.)

holtskinner replied:
Page number is populated by document ai directly, and it functions as a 1-based index for the page in the document.

I could also add methods to get the original pages/entities by their unique identifiers (page_number and id respectively)

dizcology commented on samples/snippets/table_sample.py:
```diff
@@ -35,7 +35,7 @@ def table_sample(document_path: str, output_file_prefix: str) -> None:
             # Write table to CSV file
             output_file = f"{output_file_prefix}-{page_number}-{table_number}.csv"
```
Maybe rename these `*_number` to `*_index`.

holtskinner replied:
Ah, this should be changed to use the actual `Page.page_number` instead of a locally created index.

dizcology commented on google/cloud/documentai_toolbox/wrappers/page.py:
```diff
@@ -131,7 +131,7 @@ class Block:
     """
 
-    documentai_block: documentai.Document.Page.Block
+    documentai_block: dataclasses.InitVar[documentai.Document.Page.Block]
```
Another small inconsistency is that now only some of the wrapper classes require the `shard_index` init argument, but not all.  Can we make it easy for the next engineer to tell how to decide whether a wrapper class should vs should not have `shard_index`?

holtskinner replied:
Removed these

holtskinner committed:
Addressed some code review comments

- Add incomplete shard check
- refactor some of the vision helpers to simplify imports
- Fix documentation formatting issue

holtskinner committed:
fix test errors

holtskinner committed:
Fix regex in test

holtskinner committed:
Change crop_image() return type to optional

holtskinner committed:
Update Table to use initVar

holtskinner committed:
Update table sample test

holtskinner committed:
Add post init for formfields

holtskinner committed:
Add postinit to Block, Line, Paragraph

holtskinner committed:
Update Tests

holtskinner committed:
refactor: Re-added documentai elements as ful members of dataclasses

- Avoids some backwards incompatibility issues
- Fixed bug where Entity page numbers would not line up with multi-shard documents

holtskinner committed:
Merge branch 'main' into memory

holtskinner committed:
fix: Removed referebce to removed function

holtskinner committed:
refactor: Moved more vision conversion to vision_helpers.py

holtskinner committed:
Removed extra import

dizcology commented on google/cloud/documentai_toolbox/wrappers/document.py:
```diff
@@ -582,7 +559,7 @@ def get_form_field_by_name(self, target_field: str) -> List[FormField]:
 
     def form_fields_to_dict(self) -> Dict:
-        r"""Returns Dictionary of form fields in document.
+        r"""Returns dictionary of form fields in document.
```
Make the type annotation (and maybe the docstring too) more informative.  Is the return type `Dict[str, str]` here?

holtskinner replied:
Added a more explicit type annotation `Dict[str, Union[str, List[str]]]`

dizcology commented on google/cloud/documentai_toolbox/wrappers/document.py:
```diff
@@ -767,13 +752,15 @@ def export_images(
                 continue
-
             output_filename = (
                 f"{output_file_prefix}_{index}_{entity.type_}.{output_file_extension}"
```
Update the docstring to include the output filename format.

holtskinner replied:
This is already in the docstring for the function. Am I missing something?

dizcology commented on google/cloud/documentai_toolbox/wrappers/document.py:
```diff
@@ -722,27 +699,35 @@ def split_pdf(self, pdf_path: str, output_path: str) -> List[str]:
+            self.convert_document_to_annotate_file_response()
         )
 
     def export_images(
```
The method name here seems to suggest that it exports all images, but actually only those that belong to entities?

(Perhaps all images belong to entities?)

holtskinner replied:
Note - DocAI doesn't do any general image recognition currently, it's only for specific entity extraction

dizcology commented on google/cloud/documentai_toolbox/wrappers/entity.py:
```diff
@@ -26,31 +27,46 @@
+            Required. The original `google.cloud.documentai.Document.Entity` object.
+        page_offset (int):
+            Optional. The page offset (for sharded Documents) of the `google.cloud.documentai.Document` containing
+            the `google.cloud.documentai.Document.Entity`. `page_refs.page` is relative to the shard, not the entire Document.
```
`page_refs` seems to be an attribute of the underlying documentai message and is used only in the implementation, and this part of this docstring would not help the user understand what `page_offset` means.

holtskinner replied:
Rewrote this doc string. Is this more clear?

dizcology commented on google/cloud/documentai_toolbox/wrappers/entity.py:
```diff
@@ -59,32 +75,36 @@ def __post_init__(self):
+        page_refs = self.documentai_entity.page_anchor.page_refs
+        if page_refs:
+            self.start_page = int(page_refs[0].page) + page_offset
+            self.end_page = int(page_refs[-1].page) + page_offset
```
We should add a unit test for these lines of code above.

holtskinner replied:
Done

dizcology commented on google/cloud/documentai_toolbox/wrappers/page.py:
```diff
@@ -29,15 +29,27 @@ class Table:
         documentai_table (google.cloud.documentai.Document.Page.Table):
             Required. The original google.cloud.documentai.Document.Page.Table object.
+        document_text (str):
+            Required. UTF-8 encoded text in reading order from the document.
```
Is `document_text` the text for the whole document, or just that of a page containing the object (e.g. a Table)?  If it is just the containing page's text, perhaps let's replace the `document_text` attribute with a `_page` attribute instead (here and below).

holtskinner replied:
Throughout this code, `document_text` is the text of the full documentai Document, not just the page.

holtskinner committed:
Addressed review comments

- Added a more specific type for dictionary. `Dict[str, Union[str, List[str]]]`
- Rewrote Docstring for `Entity.page_offset`
- Wrote unit test for `page_offset`

holtskinner committed:
Update tests to improve coverage

holtskinner committed:
Merge branch 'main' into memory
